src/ObjCRuntime/TransientString.cs

```diff
@@ -50,5 +50,33 @@ public void Dispose ()
 		}
 
 		public static implicit operator IntPtr (TransientString str) => str.ptr;
+
+
+		[DllImport ("/usr/lib/libc.dylib")]
+		static extern void free (IntPtr ptr);
+
+		public static string? ToStringAndFree (IntPtr ptr, Encoding encoding = Encoding.Auto)
+		{
+			string? result = null;
+			switch (encoding) {
+			case Encoding.Auto:
+				result = Marshal.PtrToStringAuto (ptr);
+				break;
+			case Encoding.BStr:
+				result = Marshal.PtrToStringBSTR (ptr);
+				break;
+			case Encoding.Ansi:
+				result = Marshal.PtrToStringAnsi (ptr);
+				break;
+			case Encoding.Unicode:
+				result = Marshal.PtrToStringUni (ptr);
+				break;
+			default:
+				throw new ArgumentOutOfRangeException (nameof (encoding));
+			}
+			if (ptr != IntPtr.Zero)
+				free (ptr);
```

Comment:
You can call Marshal.FreeHGlobal here (and no need for the IntPtr.Zero check).

Unfortunately the documentation doesn't say what Marshal.FreeHGlobal will do on non-Windows platforms, but according to the source code, Marshal.FreeHGlobal [will call NativeMemory.Free](https://github.com/dotnet/runtime/blob/7d23ca02b4403c0df32d09862ec7d6a3867e13fe/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Unix.cs#L66
), and [NativeMemory.Free](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativememory.free?view=net-7.0#system-runtime-interopservices-nativememory-free(system-void*)) is documented to call `free`: "This method is a thin wrapper over the C free API."